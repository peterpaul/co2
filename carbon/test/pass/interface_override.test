/*
 * carbon Compiler and programming language for libco2
 *
 * Copyright (C) 2011 Peterpaul Taekele Klein Haneveld
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


interface A {
	int aa (int a, int b);
	int ab (int a, int b);
}

interface B {
	int ba (int a, int b);
	int bb (int a, int b);
}

class C {
	int aa (int a, int b) return a + b;
	int bb (int a, int b) return a + b;
	int ca (int a, int b) return a + b;	
}

class D : C, A, B {
	int aa (int a, int b) return super (a, b);
	int bb (int a, int b) return super (a, b);

	int ab (int a, int b) return a + b;
	int ba (int a, int b) return a + b;

	int ca (int a, int b) return super (a, b);
	int da (int a, int b) return a + b;
}

int main () {
	D d = new D();
	A a = (A) d;
	B b = (B) d;
	C c = (C) d;
	d.aa(4, 5);
	a.aa(4, 5);
	d.ab(4, 5);
	a.ab(4, 5);
	d.ba(4, 5);
	b.ba(4, 5);
	d.bb(4, 5);
	b.bb(4, 5);
	d.ca(4, 5);
	c.ca(4, 5);
	d.da(4, 5);
	return 0;
}
