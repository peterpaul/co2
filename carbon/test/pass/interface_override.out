14
14
15
15
16
16
16
16
20
20
18
